src/Init/Data/BitVec/Bitblast.lean

```diff
@@ -429,6 +429,417 @@ theorem shiftLeft_eq_shiftLeftRec (x : BitVec w₁) (y : BitVec w₂) :
+        apply toNat_shiftConcat_lt_of_lt_of_lt_two_pow <;> bv_omega
+      rw [BitVec.toNat_sub_eq_toNat_sub_toNat_of_le hdr']
+      omega
+    case neg.hqWidth =>
+      apply toNat_shiftConcat_lt_of_lt_of_lt_two_pow <;> omega
+    case neg.hdiv =>
+      have rltd' := (BitVec.le_iff_not_lt.mp rltd)
+      simp only [qr.toNat_shiftRight_sub_one_eq h]
+      simp only [BitVec.toNat_sub_eq_toNat_sub_toNat_of_le rltd']
+      simp only [toNat_shiftConcat_eq_of_lt_of_lt_two_pow (hk := qr.wr_lt_w h) (hx := h.hrWidth)]
+      simp only [BitVec.le_def, toNat_shiftConcat_eq_of_lt_of_lt_two_pow (hk := qr.wr_lt_w h) (hx := h.hrWidth)] at rltd'
+      simp only [toNat_shiftConcat_eq_of_lt_of_lt_two_pow (hk := qr.wr_lt_w h) (hx := h.hqWidth),
+        h.hdiv, Nat.mul_add]
+      bv_omega
+
+/-! ### Core divsion algorithm circuit -/
+
+/-- A recursive definition of division for bitblasting, in terms of a shift-subtraction circuit. -/
+def divRec (w wr wn : Nat) (n d : BitVec w) (qr : DivModState w) :
+    DivModState w :=
+  match wn with
+  | 0 => qr
+  | wn + 1 =>
+    divRec w (wr + 1) wn n d <| divSubtractShift n d (wn + 1) qr
+
+@[simp]
+theorem divRec_zero (qr : DivModState w) :
+  divRec w w 0 n d qr  = qr := rfl
+
+@[simp]
+theorem divRec_succ (wn : Nat) (qr : DivModState w) :
+    divRec w wr (wn + 1) n d qr =
+      divRec w (wr + 1) wn n d
+        (divSubtractShift n d (wn + 1) qr) := rfl
+
+theorem divRec_correct {n d : BitVec w} (qr : DivModState w)
+  (h : DivModState.Lawful w wr wn n d qr) : DivModState.Lawful w w 0 n d (divRec w wr wn n d qr) := by
+  induction wn generalizing wr qr
+  case zero =>
+    unfold divRec
+    simp [← h.hwrn, h]
+  case succ wn' ih =>
+    simp only [divRec]
+    apply ih
+    apply divSubtractShiftProof (w := w)
+      (wr := wr)
+      (wn := wn' + 1)
+    exact h.toLawfulShiftSubtract
+
+/-- The result of `udiv` agrees with the result of the division recurrence. -/
+theorem udiv_eq_divRec (hd : 0#w < d) :
+    let out := divRec w 0 w n d (DivModState.init w)
+    n.udiv d = out.q := by
+  have := DivModState.Lawful.init w n d hd
+  have := divRec_correct (DivModState.init w) this
+  apply DivModState.udiv_eq_of_lawful_zero this
+
+/-- The result of `umod` agrees with the result of the division recurrence. -/
+theorem umod_eq_divRec (hd : 0#w < d) :
+    let out := divRec w 0 w n d (DivModState.init w)
+    n.umod d = out.r := by
+  have := DivModState.Lawful.init w n d hd
+  have := divRec_correct (DivModState.init w) this
+  apply DivModState.umod_eq_of_lawful_zero this
+
+@[simp]
+theorem divRec_succ' (wn : Nat) (qr : DivModState w) :
+    divRec w wr (wn + 1) n d qr =
+    let r' := shiftConcat qr.r (n.getLsb wn)
+    let input : DivModState w :=
+      if r' < d then ⟨qr.q.shiftConcat false, r'⟩ else ⟨qr.q.shiftConcat true, r' - d⟩
+    divRec w (wr + 1) wn n d input := by
+    simp only [divRec_succ, divSubtractShift, Nat.add_one_sub_one, decide_eq_true_eq]
```

Comment:
Terminal simp-only? Is this necessary or should this just be simp?

Reply:
Cleaned, thanks.